Support contains for range

synapticarbors opened the pull request:
This addresses #4109 based on the python code in:
https://github.com/python/cpython/blob/master/Objects/rangeobject.c#L337-L379
.
There are a few issues that I'm unsure of how to handle:

- When testing with an imaginary `val`, I get the following error (I removed the test case, but will add it back in later):

```python
======================================================================
ERROR: test_range_contains (numba.tests.test_range.TestRange)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xyz/Projects/github/numba/numba/tests/test_range.py", line 206, in test_range_contains
    self.assertEqual(cfunc(val, *arg), pyfunc(val, *arg))
  File "/Users/xyz/Projects/github/numba/numba/dispatcher.py", line 351, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/xyz/Projects/github/numba/numba/dispatcher.py", line 318, in error_rewrite
    reraise(type(e), e, None)
  File "/Users/xyz/Projects/github/numba/numba/six.py", line 658, in reraise
    raise value.with_traceback(tb)
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<built-in function contains>) with argument(s) of type(s): (range_state_int64, complex128)
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
In definition 2:
    AttributeError: 'Complex' object has no attribute 'imag'
    raised from /Users/xyz/Projects/github/numba/numba/targets/rangeobj.py:260
In definition 3:
    AttributeError: 'Complex' object has no attribute 'imag'
    raised from /Users/xyz/Projects/github/numba/numba/targets/rangeobj.py:260
In definition 4:
    All templates rejected with literals.
In definition 5:
    All templates rejected without literals.
In definition 6:
    All templates rejected with literals.
In definition 7:
    All templates rejected without literals.
In definition 8:
    All templates rejected with literals.
In definition 9:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: typing of intrinsic-call at /Users/xyz/Projects/github/numba/numba/tests/test_range.py (72)

File "numba/tests/test_range.py", line 72:
def range_contains(val, start, stop, step):
    <source elided>
    for r in (r1, r2, r3):
        tmp.append(val in r)
        ^
```

It seems like `val` should have an `.imag` attribute, but maybe I'm missing something.

- There is some behavior around float in range, imaginary in range, and non-numerical object in range, that python displays, that I'm trying to capture correctly. 
    - Floats e.g `1.0 in range(10) is True`, but any non-zero decimal value returns `False`. 
    - Complex with non-zero imaginary parts always return False, but if the imaginary part is zero, it evaluate like a standard `int`
    - For any non-numerical object, `x in range(N)` appears to always return `False`, but it looks like from the python C-code, they actually default to a full iteration search using `_PySequence_IterSearch`, which is born out by running `timeit` on something like `{'a': 1} in range(N)`, varying `N`. I'm just automatically returning `False`, but maybe I'm missing a case. I'm not sure why CPython is executing a search since you know all of the items in `range(N)` are integers.

    I think I've captured all of the edge cases, but I'm not 100% sure

esc commented:
@synapticarbors thanks again for using Numba and contributing this, it is much appreciated.

Witghout having looked at the cpython source, It sounds like Python is doing some sort of 'cast' or an attempt to cast the value to an integer. I did checkout the current state of the PR and I think the way you are handling reals right now might be a good approach and you could probably handle complex values like this too.

The reason why your code is failing on the `imag` attribute is due to the fact that in this scope, you only have the Numba type object, but not the actual value. Checking the value of the `imag` part is only possible within an implementation function. If you aren't aware of it already, you can look at the docs in https://github.com/numba/numba/pull/3951 which should help to understand this different scopes when using `@overload`.

esc commented:
Here is an alternative implementation that you could use, note that I have just quickly written this down and it is likely to be imperfect but should illustrate a general idea for an implementation structure:

```
@register_jitable
def impl_contains_helper(robj, val):
    if robj.step > 0 and (val < robj.start or val >= robj.stop):
        return False
    elif robj.step < 0 and (val <= robj.stop or val > robj.start):
        return False

    return ((val - robj.start) % robj.step) == 0

@overload(operator.contains)
def impl_contains(robj, val):

    def impl_false(robj, val):
        return False

    if not isinstance(robj, types.RangeType):
        return

    elif not isinstance(val, types.Number):
        return impl_false

    elif isinstance(val, types.Integer):
        return impl_contains_helper

    elif isinstance(val, types.Float):
        def impl(robj, val):
            if val.imag != 0:
                return False
            else:
                return impl_contains_helper(robj, int(val))
        return impl

    elif isinstance(val, types.Complex):
        def impl(robj, val):
            if val % 1 != 0:
                return False
            else:
                return impl_contains_helper(robj, int(val))
        return impl
```

esc commented:
@synapticarbors I made a small mistake above, the code for `Float`  and `Complex` should be swapped.

esc commented:
@synapticarbors here is a version that is likely to work for you:

```
    elif isinstance(val, types.Float):
        def impl(robj, val):
            if val % 1 != 0:
                return False
            else:
                return impl_contains_helper(robj, int(val))
        return impl

    elif isinstance(val, types.Complex):
        def impl(robj, val):
            if val.imag != 0:
                return False
            else:
                return impl_contains_helper(robj, int(val.real))
        return impl
```

synapticarbors commented:
@esc -- Thanks for the suggestions. I basically used your suggestion with the small addition of catching a case under the `Complex` branch that wasn't handled, as well as adding proper handling of booleans vals. 

Please have a look at:
- The use of `njit` and `jit(..., forceobj=True)` in the tests. Throughout the various tests, in some places this type of simple jitting is used, but in others a more complex harness is employed. I'm not sure when which approach is appropriate.

- In the CPython implementation, it looks like non-numbers cause a full iteration over the range object. Here I'm just uniformly saying that it should return `False`. I'm not sure why the CPython version works the way that it does, but am I possibly missing some case that isn't entirely obvious?

esc commented:
@synapticarbors thanks for taking this further. I saw that you force-pushed the branch and this is highly discouraged in the Numba project because Github doesn't handle this gracefully and we risk loosing existing reviews and comments. So, sorry about that and thanks! I'll leave a few more comments in the code regarding specific issues that I would recommend you address.

Regarding the two questions.

* `njit` (or `jit(nopython=True)`) should suffice. I am not sure why you would need `forceobj=True`. I believe the test will also work without it.

* I thought about this for quite some time and I can not imagine a scenario where not rejecting immediately would make sense.

esc commented on numba/targets/rangeobj.py:
```diff
@@ -230,5 +232,48 @@ def get(rnge):
+                return impl_contains_helper(robj, int(val))
+        return impl
+
+    elif isinstance(val, types.Complex):
```
You might be able to get away with merging the `Float`  and `Complex` cases, but it might have a performance impact. Like this it contains some duplicated code, on the other hand merging the two might  make it slower for `Floats` as there is an additional check involved. Only `@timeit` will tell.

synapticarbors replied:
I'm just going to leave them separate for now unless there is a strong preference for combining them.

esc replied:
o.k.

esc commented on numba/tests/test_range.py:
```diff
@@ -177,6 +187,31 @@ def test_range_attrs(self):
+                   (-1, 4, 10),
+                   (5, -5, -2),]
+
+        vallist = [-10, -6, -5, -4, -2, -1, 0,
```
It would be good to split the `vallist` into several for each type to make it easier to read. You can then chain them together. This will make it clearer what values are being tested in what type category.

```
bool_vals = ...
int_vals = ...
float_vals = ...
complex_vals = ...
non_numeric_vals = ...
for val in int_vals + float_vals ....:
    pass
```

synapticarbors replied:
Done

esc commented on numba/tests/test_range.py:
```diff
@@ -177,6 +187,31 @@ def test_range_attrs(self):
+            for val in vallist:
+                self.assertEqual(cfunc(val, *arg), pyfunc(val, *arg))
+
+        cfunc_obj = jit(pyfunc, forceobj=True)
```
Yeah, not sure if you really need `forceobj=True` here.

esc replied:
If not, you can add the objects to the loops above.

synapticarbors replied:
Attempting to pass a `dict` into the `nopython` jitted functions causes an error since Numba does not know how to type the `dict` in `nopython` mode. I put `forceobj`there just to make it explicit, and given future changes to numba:

http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit

esc replied:
sounds good.

esc commented on numba/tests/test_range.py:
```diff
@@ -177,6 +187,31 @@ def test_range_attrs(self):
+    @tag('important')
+    def test_range_contains(self):
+        pyfunc = range_contains
+        arglist = [(0, 0, 1),
```
You may want to check all permutations of `0, 1, -1` to make sure everything is covered. Also, you may want to consider looking at max values too to check not only around 0 and 1 but also at the 'fringes' of the number line.

synapticarbors replied:
I added all of the permutations mentioned. I wasn't sure about the need for the max value checks, so I left them out for now.

esc commented on numba/tests/test_range.py:
```diff
@@ -63,6 +63,16 @@ def range_attrs(start, stop, step):
+    r1 = range(start)
+    r2 = range(start, stop)
+    r3 = range(start, stop, step)
+    tmp = []
```
It's more pythonic to return a list comprehension here.

```
return [val in r for r in (r1, r2, r3)]
```

synapticarbors replied:
done

synapticarbors commented:
@esc - thanks for letting me know about the force push policy. I was using `git commit --amend` to try to limit the number of small commits in the PR, but I'll discontinue that going forward. I think I addressed most of your concerns. See individual comments to your inline points.

esc commented:
@synapticarbors this is looking good. The Windows 2.7 test is failing for another reason. There is a PR that will fix this:

https://github.com/numba/numba/pull/4138#pullrequestreview-245291204

Hopefully this will be merged to master soon.

synapticarbors commented:
@esc I pulled in upstream to get that fix, so hopefully everything will come up green when the CI runs.

sklam commented:
I thk it's ready to merge once CI passes.
